PR 2634 CI - Fix the tool_choice format for named choice by adapting OpenAIs scheme

drbh opened the pull request:
This PR re opens #2634 to run CI

drbh commented:
@linusbierhoff thank you for the contribution! 🙏 I've opened this PR to run CI, and add some other tool calling improvements along with your changes.

This PR now removes the old `ToolType` in favor of a more simple `ChatCompletionToolChoiceOption`. This new struct includes the addition of `TypedChoice` and correctly allows a user to specify the following tools as strings or as an object. Additionally this PR includes support for the `required` tool_choice, which forces the model to use one of the user provided tools. Lastly the new struct also improves the openapi spec and makes the tool choice a `oneOf` which improves the spec readability

Narsil commented on router/src/lib.rs:
```diff
@@ -956,48 +966,72 @@ pub fn default_tool_prompt() -> String {
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema, Default)]
+#[serde(from = "ToolTypeDeserializer")]
+pub enum ChatCompletionToolChoiceOption {
```
Can we avoid java naming please ?

drbh replied:
yea I can revert the name change if thats better.

The name change is to align with the naming that OpenAI is using in their spec. https://github.com/openai/openai-openapi/blob/d033c364c6574068ee89f3d5f845a4830bddd503/openapi.yaml#L9581, I aligned the names because I thought it may be helpful in automatically checking the spec, however it is a very long name..

Narsil replied:
> I thought it may be helpful in automatically checking the spec

Then let's wait for it to be actually relevant, it's not usually a good thing to do things "because we might need it". Either you need it or you don't.

Personally I don't think it has any relevance to use their names.

drbh replied:
got it that make sense, I've changed the name to `ToolChoice` in the latest commit. (its much cleaner 👍)

Narsil commented on router/src/lib.rs:
```diff
@@ -956,48 +966,72 @@ pub fn default_tool_prompt() -> String {
+    #[default]
     NoTool,
+    /// Means the model must call one or more tools.
+    #[schema(rename = "required")]
```
`schema rename` is not what you want, you want serde rename most likely.

drbh replied:
oh I actually included it to improve the generated schema

this macro changes the `enum` on the schema to show the expected lowercase text instead of `Required`
```json
{
  "type": "string",
  "description": "Means the model must call one or more tools.",
  "enum": [
    "required"
  ]
},
```

Narsil replied:
Yes, and serde(renamed) actually modifies what the input is from the user.

Narsil replied:
Actually a better serde(rename) on the whole struct should fix every field (both schema and actual JSON parsing)

drbh replied:
oh yea thats much better, updated in the latest commit

Narsil commented on router/src/lib.rs:
```diff
@@ -956,48 +966,72 @@ pub fn default_tool_prompt() -> String {
+    Required,
     /// Forces the model to call a specific tool.
     #[schema(rename = "function")]
+    #[serde(alias = "function")]
```
Can we have link in rustdoc to where all these attributes are explained, possibly in the original openai spec (or anthropic)

drbh replied:
yep, just added a useful link above the enum that explains all 4 of the options in OpenAI's docs this link will also be included in the openapi schema for the `ChatCompletionToolChoiceOption` description

Narsil commented on router/src/lib.rs:
```diff
@@ -1604,4 +1638,40 @@ mod tests {
+
+        let named = r#"{"tool_choice":"myfn"}"#;
+        let de_named: TestRequest = serde_json::from_str(named).unwrap();
+        assert_eq!(de_named.tool_choice, ref_choice);
```
```suggestion
        assert_eq!(de_named.tool_choice,ChatCompletionToolChoiceOption::Function(FunctionName {
            name: "myfn".to_string(),
        }) );
```

Please avoid indirection in tests. Tests are about being as dumb as possible, so a reader can easily assess if the test is correct, or outdated. Every indirection makes that harder to reason about.

drbh replied:
great points, indirection removed in the latest commit

Narsil commented on router/src/infer/tool_grammar.rs:
```diff
@@ -20,44 +20,47 @@ impl ToolGrammar {
-        tools.push(no_tool);
+        // add the no_tool function to the tools as long as we are not required to use a specific tool
+        if tool_choice != ChatCompletionToolChoiceOption::Required {
+            let no_tool = Tool {
```
Isn't this only ever used in the `auto` branch ? Wouldn't putting it in the match branch make more sense ?

drbh replied:
yep good point. Just moved into the match branch and removed the `let mut tools` and only mutate tools if the Auto option is chosen. thanks

Narsil commented on router/src/infer/tool_grammar.rs:
```diff
@@ -20,44 +20,47 @@ impl ToolGrammar {
-            ToolType::NoTool => return Ok((tools, None)),
+            ChatCompletionToolChoiceOption::Required => tools.clone(),
+            ChatCompletionToolChoiceOption::Auto => tools.clone(),
+            ChatCompletionToolChoiceOption::NoTool => return Ok((tools, None)),
```
Why return some tools, if we don't want to use them ? Wouldn't discarding them be better ?

drbh replied:
this was originally added to allow passing tools to the chat template without enforcing a grammar, which was introduced in this PR https://github.com/huggingface/text-generation-inference/pull/2454, but I believe should be removed in favor of discarding the tools. The latest commits discard tools when the choice is `NoTool`

Narsil commented on router/src/infer/tool_grammar.rs:
```diff
@@ -20,44 +20,47 @@ impl ToolGrammar {
-            ToolType::OneOf => tools.clone(),
-            ToolType::NoTool => return Ok((tools, None)),
+            ChatCompletionToolChoiceOption::Required => tools.clone(),
+            ChatCompletionToolChoiceOption::Auto => tools.clone(),
```
Why is this `.clone` needed ? I know it was there before, but since we're upgrading this, couldn't we get rid of it ?

drbh replied:
yes agreed they should be removed. Just removed all of the tool clones above and improved the match in general

Narsil commented on router/src/infer/tool_grammar.rs:
```diff
@@ -20,44 +20,47 @@ impl ToolGrammar {
-            ToolType::NoTool => return Ok((tools, None)),
+            ChatCompletionToolChoiceOption::Required => tools.clone(),
+            ChatCompletionToolChoiceOption::Auto => tools.clone(),
+            ChatCompletionToolChoiceOption::NoTool => return Ok((tools, None)),
```
Also I think early returns in `match` branches are bug nests, it's easy to overlook those branches, returning an empty slice and letting the rest roll seems slightly better (code reading wise)

drbh replied:
good point, updated in the latest to commit to discard the tools and continue executing the function normally

Narsil commented on router/src/lib.rs:
```diff
@@ -849,8 +849,8 @@ pub(crate) struct ChatRequest {
-    #[schema(nullable = true, example = "null")]
-    pub tool_choice: ToolChoice,
+    #[schema(nullable = true, default = "null", example = "null")]
+    pub tool_choice: Option<ChatCompletionToolChoiceOption>,
```
Why the change to `Option` you have the `NoTool` choice, there shouldn't be a need to `Option` wrapping is there ?

drbh replied:
`Option` is needed so the user can send a http request without a `tool_choice` param included. If `tool_choice` is `None` (user did not include a tool_choice in the request), `ChatCompletionToolChoiceOption` defaults to `NoTool`

Narsil replied:
Doesn't a default take care of that ? And shouldn't the default be "auto"?

Narsil commented on router/src/lib.rs:
```diff
@@ -906,8 +906,18 @@ impl ChatRequest {
             response_format,
-            tools,
-            tool_choice,
+            tools.clone(),
```
No clone.

drbh replied:
removed, it was previously required because of a check of the `tools` vec that was below this line (moved it up and removed the clone). thanks

Narsil commented on router/src/lib.rs:
```diff
@@ -906,8 +906,18 @@ impl ChatRequest {
-            tool_choice,
+            tools.clone(),
+            // unwrap or default (use "auto" if tools are present, and "none" if not)
+            tool_choice.map_or_else(
```
That should be already done in the deserialization, there are about 3 different function calls to `preprare_chat_input`.

Narsil commented:
`required` is that an OpenAI spec ?

Can you point to the original references for the things we're adding.
Also dropping the old code is a pretty bad breaking change, it should be in big bold red !.

Why are we dropping support for the old way, if it was there, is it documented anywhere ? Should we update the old doc ? 

This PR is also seriously lacking integration tests for all of these. The deserialization tests are nice, but not enough.

drbh commented:
> required is that an OpenAI spec ?

yea `required` is supported by OpenAI, see docs here: https://platform.openai.com/docs/guides/function-calling/configuring-function-calling-behavior-using-the-tool_choice-parameter. Note** OpenAI added this parameter after TGI's initial tool implementation (hence why it was missing), additionally prior to https://github.com/huggingface/text-generation-inference/pull/2614 `auto` and `required` had the same functionality in TGI, however now its more pressing/great to have the api's aligned.

> Also dropping the old code is a pretty bad breaking change, it should be in big bold red !.

yea regarding the removed code - as the original PR noted, the `function` tool choice was expecting the incorrect format. I believe I confused some of the function calling formats and ended up adding one that was not aligned with any major LLM api. 

regarding flagging the breaking changes, would it be best to include the old API but throw a helpful error if it is used? Currently I've removed the path in favor of the corrected one. 

> Why are we dropping support for the old way, if it was there, is it documented anywhere ? Should we update the old doc ?

we're dropping support since it was incorrectly implemented, we do not have documentation specific to this tool choice, however I'm going to push another commit with updated docs for the supported tool choice types

> This PR is also seriously lacking integration tests for all of these. The deserialization tests are nice, but not enough.

just added a test for `required` and one for a function object in `test_tools_llama.py`

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/text-generation-inference/pr_2645). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

Narsil commented on integration-tests/models/test_tools_llama.py:
```diff
@@ -327,3 +329,102 @@ async def test_flash_llama_grammar_tools_sea_creatures_stream(
+                last_response = response
+
+    assert count == 30
+    print(tool_calls_generated)
```
Oops

drbh replied:
oops 😬, removed in latest commit. thanks!

Narsil commented on router/src/lib.rs:
```diff
@@ -903,15 +903,47 @@ impl ChatRequest {
+        // unwrap or default (use "auto" if tools are present, and "none" if not)
+        let tool_choice = tool_choice.unwrap_or_else(|| {
+            if tools.is_some() {
+                ChatCompletionToolChoiceOption::Auto
+            } else {
+                ChatCompletionToolChoiceOption::NoTool
+            }
```
What happens if tool_choice is "required" and no tools are sent ?

("auto", tools = []) we should just ignore the grammar right ?

drbh commented on integration-tests/models/test_tools_llama.py:
```diff
@@ -327,3 +329,102 @@ async def test_flash_llama_grammar_tools_sea_creatures_stream(
+            "tools": tools,
+            "tool_choice": {
+                "type": "function",
+                "function": {"name": "get_current_weather"},
```
agreed, just updated to choose `get_n_day_weather_forecast` instead

Narsil commented on router/src/infer/tool_grammar.rs:
```diff
@@ -20,19 +19,19 @@ impl ToolGrammar {
-        // if no tools are provided, we return None
+        // if no tools are provided, we return None and an empty vec
         if tools.is_empty() {
-            return Ok((tools, None));
+            return Ok((Vec::with_capacity(0), None));
```
`vec![]` ?

drbh replied:
updated in latest commit

Narsil replied:
I don't see it.

drbh replied:
`let tools_to_use = match tool_choice` was updated to return a `vec![]` rather than a `Vec::with_capacity(0)`, and the conditional was moved below `tools_to_use` to return none in both cases (no tools supplied, or if NoTools specified) https://github.com/huggingface/text-generation-inference/blob/16e4c38898ed36ee44237314ce823bf3d81ea90c/router/src/infer/tool_grammar.rs#L55

Narsil commented on router/src/lib.rs:
```diff
@@ -1002,21 +996,18 @@ pub struct FunctionName {
     /// Means the model will not call any tool and instead generates a message.
     #[schema(rename = "none")]
-    #[default]
     NoTool,
```
Isn't this `no_tool` with `snake_case`  ?This should mean a rename of this property or `None`,  no ?

I don't think `schema(rename)` imples a serde `rename`.

drbh replied:
good catch, updated to `serde(rename)` and added a test to ensure that the `"none"` values is respected correctly

drbh commented on router/src/lib.rs:
```diff
@@ -850,13 +850,13 @@ pub(crate) struct ChatRequest {
     #[serde(default)]
     #[schema(nullable = true, default = "null", example = "null")]
-    pub tool_choice: Option<ChatCompletionToolChoiceOption>,
+    pub tool_choice: Option<ToolChoice>,
```
oh yes that worked, I was originally under the impression the Option was still needed to handle all http request but this is much cleaner. Thank you

drbh commented on router/src/lib.rs:
```diff
@@ -850,13 +850,13 @@ pub(crate) struct ChatRequest {
     /// NOTE: A request can use `response_format` OR `tools` but not both.
     #[serde(default)]
-    #[schema(nullable = true, default = "null", example = "null")]
+    #[schema(nullable = true, default = "auto", example = "auto")]
```
yes thank you, updated the correct argument in the latest commit

drbh commented:
merging optimistically as CI is green and issues have been resolved. Also mentioned in the internal #text-generation-inference Slack channel.

Will watch for any regressions and revert if needed
